include/pagmo/algorithms/pso.hpp

```diff
@@ -0,0 +1,875 @@
+                    neighb[pidx].push_back(pidx - j + swarm_size);
+                } else {
+                    neighb[pidx].push_back(pidx - j);
+                }
+            }
+            for (decltype(radius) j = 1u; j <= radius; j++) {
+                if (pidx + j >= swarm_size) {
+                    neighb[pidx].push_back(pidx + j - swarm_size);
+                } else {
+                    neighb[pidx].push_back(pidx + j);
+                }
+            }
+        }
+    }
+
+    /*! @brief Von Neumann neighborhood
+     *  (increments on particles' lattice coordinates that produce the coordinates of their neighbors)
+     *
+     *  The von Neumann neighbourhood of a point includes all the points at a Hamming distance of 1.
+     *
+     *  - http://en.wikipedia.org/wiki/Von_Neumann_neighborhood
+     *  - http://mathworld.wolfram.com/vonNeumannNeighborhood.html
+     *  - http://en.wikibooks.org/wiki/Cellular_Automata/Neighborhood
+     */
+    const int vonNeumann_neighb_diff[4][2] = {{-1, 0}, {1, 0}, {0, -1}, {0, 1}};
+
+    /**
+     *  @brief Arranges particles in a lattice, where each interacts with its immediate 4 neighbors to the N, S, E and
+     * W.
+     *
+     *  ''The population is arranged in a rectangular matrix, for instance, 5 x 4
+     *  in a population of 20 individuals, and each individual is connected to
+     *  the individuals above, below and on both of its sides, wrapping the edges'' \n
+     *  [Kennedy and Mendes, 2006] http://dx.doi.org/10.1109/TSMCC.2006.875410
+     *
+     *  ''Given a population of size n, the von Neumann neighborhood was
+     *  configured into r rows and c columns, where r is the smallest integer
+     *  less than or equal to sqrt(n) that evenly divides n and c = n / r'' \n
+     *  [Mohais et al., 2005] http://dx.doi.org/10.1007/11589990_80 \n
+     *  (there's an error in the description above: "smallest integer" should
+     *  instead be "highest integer")
+     *
+     *  @param[out] neighb definition of the swarm's topology
+     */
+    void initialize_topology__von(std::vector<std::vector<vector_double::size_type>> &neighb) const
+    {
+        int swarm_size = static_cast<int>(neighb.size());
+        int cols, rows; // lattice structure
+        int p_x, p_y;   // particle's coordinates in the lattice
+        int n_x, n_y;   // particle neighbor's coordinates in the lattice
+
+        rows = static_cast<int>(std::sqrt(swarm_size));
+        while (swarm_size % rows != 0u)
+            rows -= 1;
+        cols = swarm_size / rows;
+
+        for (decltype(swarm_size) pidx = 0u; pidx < swarm_size; pidx++) {
+            p_x = pidx % cols;
+            p_y = pidx / cols;
+
+            for (unsigned int nidx = 0u; nidx < 4u; nidx++) {
+                n_x = (p_x + vonNeumann_neighb_diff[nidx][0]) % cols;
+                if (n_x < 0)
+                    n_x = cols + n_x; // sign of remainder(%) in a division when at least one of the operands is
```

Comment:
This comment is not true anymore as of C++11, the sign in a modulo operation is the sign of the dividend.

http://stackoverflow.com/questions/13100711/operator-modulo-change-in-c-11